ray-operator/controllers/ray/expectations/active_expectation.go

```diff
@@ -0,0 +1,382 @@
+		if err := ae.items.Add(item); err != nil {
+			return err
+		}
+
+		return nil
+	}
+
+	item.(*ActiveExpectationItem).Action = action
+	item.(*ActiveExpectationItem).RecordTimestamp = time.Now()
+	return nil
+}
+
+func (ae *ActiveExpectation) expectUpdate(kind ExpectedResourceType, name string, resourceVersion int64) error {
+	key := fmt.Sprintf("%s/%s", kind, name)
+	item, exist, err := ae.items.GetByKey(key)
+	if err != nil {
+		return fmt.Errorf("fail to get active expectation item for %s when expecting: %s", key, err)
+	}
+
+	ae.recordTimestamp = time.Now()
+	if !exist {
+		item = &ActiveExpectationItem{Client: ae.Client, Name: name, Kind: kind, Key: key, Action: Update, ResourceVersion: resourceVersion, RecordTimestamp: time.Now()}
+		if err := ae.items.Add(item); err != nil {
+			return err
+		}
+
+		return nil
+	}
+
+	ea := item.(*ActiveExpectationItem)
+	ea.Action = Update
+	ea.ResourceVersion = resourceVersion
+	ea.RecordTimestamp = time.Now()
+
+	return nil
+}
+
+func (ae *ActiveExpectation) isSatisfied() (satisfied bool, err error) {
+	items := ae.items.List()
+
+	satisfied = true
+	for _, item := range items {
+		itemSatisfied, itemErr := func() (satisfied bool, err error) {
+			defer func() {
+				if satisfied {
+					ae.items.Delete(item)
+				} else if ae.recordTimestamp.Add(ExpectationsTimeout).Before(time.Now()) {
+					panic("expected panic for active expectation")
+				}
+			}()
+
+			satisfied, err = item.(*ActiveExpectationItem).isSatisfied(ae.namespace)
+			if err != nil {
+				return false, err
+			}
+
+			return satisfied, nil
+		}()
+
+		if itemErr != nil && err == nil {
+			err = fmt.Errorf("fail to check satisfication for subject %s, item %s: %s", ae.key, item.(*ActiveExpectationItem).Key, itemErr)
+		}
+
+		satisfied = satisfied && itemSatisfied
+	}
+
+	return satisfied, err
+}
+
+func (ae *ActiveExpectation) delete(kind, name string) error {
+	key := fmt.Sprintf("%s/%s", kind, name)
+	item, exist, err := ae.items.GetByKey(key)
+	if err != nil {
+		return fmt.Errorf("fail to delete active expectation item for %s: %s", key, err)
+	}
+
+	if !exist {
+		return nil
+	}
+
+	if err := ae.items.Delete(item); err != nil {
+		return fmt.Errorf("fail to do delete active expectation item for %s: %s", key, err)
+	}
+
+	return nil
+}
+
+type ActiveExpectationItem struct {
+	client.Client
+
+	Key             string
+	Name            string
+	Kind            ExpectedResourceType
+	Action          ActiveExpectationAction
+	ResourceVersion int64
+	RecordTimestamp time.Time
+}
+
+func (i *ActiveExpectationItem) isSatisfied(namespace string) (bool, error) {
+	switch i.Action {
+	case Create:
+		resource := ResourceInitializers[i.Kind]()
+		if err := i.Get(context.TODO(), types.NamespacedName{Namespace: namespace, Name: i.Name}, resource); err == nil {
+			return true, nil
+		} else if errors.IsNotFound(err) && i.RecordTimestamp.Add(30*time.Second).Before(time.Now()) {
```

Comment:
What does this mean? Do you mean:

(1) The Pod is not found in the informer cache.
(2) KubeRay has already submitted a `Create` request to the K8s API server at `t=RecordTimestamp`. If the `Create` request was made more than 30 seconds ago, we assume it satisfies the expectation.

I can't understand (2). If we sent a request 30 seconds ago and the informer still hasn't received information about the Pod, there are two possibilities:

* (a) There are delays between the K8s API server and the informer cache.
* (b) The creation failed.

For case (a), it is OK for the function to say that the expectation is satisfied. However, for case (b), what will happen if the creation fails and we tell the KubeRay operator it is satisfied?

Reply:
Case(b) may not occur here? Because only after the pod is successfully created, the pod is expected to be in the cache.

```go
if err := r.Create(ctx, &pod); err != nil {
	return err
}
rayClusterExpectation.ExpectCreateHeadPod(key, pod.Namespace, pod.Name)
```